### ISSUE TITLE
Incorrect guests number counting

### ISSUE THREAD

**Issue by yuliu** [mybb/mybb#3852]
### Total number of online users
Seems happening to 1.8.22. Mentioned in [Discord chat](https://discordapp.com/channels/215876847634743296/215876989733568512/661567597287964701).

The guest count here is counting spiders/bots numbers and guest numbers, with the WHERE condition `uid = 0 AND time > $timesearch`:
https://github.com/mybb/mybb/blob/mybb_1822/index.php#L64-L89

And the total count here is the summation of member count & guest count & bot count:
https://github.com/mybb/mybb/blob/mybb_1822/index.php#L187

### Number of users viewing a specific forum
Another issue related here is that if `showforumviewing` is on, the number of users viewing a specific forum is doubled for spiders/bots:
https://github.com/mybb/mybb/blob/mybb_1822/index.php#L161

## The following issue will be tracked in a separated one.

~The number of guests that are printed on online page is not equal to the count of guests shown on the Index page's statistics~

~E.g., the MyBB community forum's online page shows `Pages (11): 1 2 3 4 5 … 11 Next » Jump to page  `. The total number of pages is 11 and per page count is 20, thus the online count of all members + bots + guests should be less than or equal to 11*20=220.~

~But at this time, the index page says `401 users active in the past 15 minutes (7 members, 1 of whom is invisible, and 391 guests).`.*~

**Comment by Eldenroot** [comment]
Yes, this is quite weird and bots should not be counted.

**Comment by yuliu** [comment]
Add an issue on recounting number of users viewing a specific forum on Index page.

**Comment by dvz** [comment]
There is a missing `SUBSTR(sid,4,1) != '='` condition in guest count queries. We could replace it with a more optimal spider ID column.

[The `GROUP` query](https://github.com/mybb/mybb/blob/mybb_1822/index.php#L66-L73) may count same IPs multiple times if they're grouped into multiple locations (`location1`) - we could calculate the global count with the [simple query](https://github.com/mybb/mybb/blob/mybb_1822/index.php#L87) (leaving possible duplicates in forum view counts) to make it consistent with how it's done in `portal.php` and `online.php`. This would avoid queries that are "random" (e.g. using `MIN()` to pick a specific IP row, making forum counts inaccurate) or more expensive (partitioning to select ~most recent IP row).

**Comment by yuliu** [comment]
Oh well.. I was thinking that guests (including bots) are counted together for the total "guests" number and x viewing forum 🤣. And of course in my opinion, a bot is a guest..

Yes you're right. I neglected the guests count for x viewing forum is also grouped on the Index page, which should be the same case I've posted in https://github.com/mybb/mybb/issues/3899#issuecomment-593380837

Well then, what's the idea to correct this issue? Use the first query for counting guests (without bots) and the second query for users and spiders/bots?

**Comment by dvz** [comment]
> And of course in my opinion, a bot is a guest..

Looks like previously the bot count contributed to the total number, but wasn't included in the guest count. It would be a minor change that can always be applied, as long as it's consistent throughout MyBB.

> the guests count for x viewing forum is also grouped on the Index page, which should be the same case I've posted in #3899

If the problem in #3899 is guests being missed, it's the opposite here - guests with the same IP address, but browsing different sections, will be counted multiple times instead of only once.

It looks like re-using results from this problematic query for the global count results in discrepancies, in addition to [guest count queries](https://github.com/mybb/mybb/blob/mybb_1822/index.php#L64-L89) that mistakenly include bots (both result in the total being inflated, compared to `online.php`).
If we use an additional query (https://github.com/mybb/mybb/blob/mybb_1822/index.php#L87 but excluding bots) that counts unique IPs regardless of locations, the total should be correct.

Your PR may have the same effect as `SUBSTR(sid,4,1) != '='` - I suggest modifying the guests query instead, as the next one with its the loop was intended to handle users and bots (which should have unique entries per user/bot). I see that it addresses a _Notice_ error, though.

**Comment by yuliu** [comment]
> Looks like previously the bot count contributed to the total number, but wasn't included in the guest count. It would be a minor change that can always be applied, as long as it's consistent throughout MyBB.

Yep, it's just what I've demanded for, to make a standard or development guideline. Though you've said we can dive into the code for concepts or variable using, there may still be lots of misunderstanding.

> If the problem in #3899 is guests being missed, it's the opposite here - guests with the same IP address, but browsing different sections, will be counted multiple times instead of only once.
> It looks like re-using results from this problematic query for the global count results in discrepancies, in addition to [guest count queries](https://github.com/mybb/mybb/blob/mybb_1822/index.php#L64-L89) that mistakenly include bots (both result in the total being inflated, compared to `online.php`).

Since the session control changed to allow multiple guest sessions with same IPs (#3837), I think it's better to distinguish them as individual online guests. Yes, current code and my PR on Index page, as well as the Who's Online page are not consistent here about **guests** counting. So I suggest:
1. Each of them contributes to the total online number of guests, them the total online number.
2. Each of them has their own `location` info and contributes to corresponding "x viewing forum" respectively.
3. Above two should apply to both the Index page (this issue) and Who's Online page (#3899), which maintains consistency throughout MyBB.

> Yes, your PR may have the same effect as `SUBSTR(sid,4,1) != '='` - I suggest modifying the guests query instead, as the next one with its the loop was intended to handle users and bots (which should have unique entries per user/bot). I see that it addresses a _Notice_ error, though.

About my PR of #3900, I was following 1.8.22's code where I bare the mind that a bot is a guest. I was trying to add an extra query for registered users, bots and guests, but halted owing to unsure query performance increase. 

By the way, what's the notice error?

**Comment by dvz** [comment]
I imagine administrators would see uncertain high numbers as a nuisance, especially if the incorrect state is saved as _most users online_. Currently unrecognized spiders (that don't handle cookies) and generally browsers with cookies disabled (usually for privacy reasons nowadays) may result in hundreds of rows being created and counted as sessions, and that doesn't include malicious behavior.
This may be less of a problem on `online.php`, but but missing guests due to duplicate IPs is probably a better bet than allowing dubious large numbers in total counts (session persistence problems aside).

> By the way, what's the notice error?

Right now `$forum_viewers[$location['location1']]` is being incremented without initializing the `$location['location1']` key, resulting in _Notice: Undefined index_ ([currently hidden by MyBB](https://github.com/mybb/mybb/blob/mybb_1822/inc/class_error.php#L99-L104)).

**Comment by yuliu** [comment]
> I imagine administrators would see uncertain high numbers as a nuisance, especially if the incorrect state is saved as _most users online_. Currently unrecognized spiders (that don't handle cookies) and generally browsers with cookies disabled (usually for privacy reasons nowadays) may result in hundreds of rows being created and counted as sessions, and that doesn't include malicious behavior.

That's kind of untrue based on my recent observations on my own forum.

> This may be less of a problem on `online.php`, but but missing guests due to duplicate IPs is probably a better bet than allowing dubious large numbers in total counts (session persistence problems aside).

Well, let me make my thoughts clear as follows:
- Spiders/bots nowadays usually utilize distributed computing, i.e., come from multiple IP addresses, to access and index webpages to avoid using single IP which will simply be banned by website.
  - That means that once we can identify them as regular spiders/bots, they won't contribute too much to the dubious high online number record. Anyway even if we don't mark them as identified spiders/bots, they still come to visit the forum from multiple IPs, aren't they still counted as real users that would probably contribute to the _most users online_?
- Some bad bots even don't have a clear UA for us to effectively mark them and most of them are also using multiple IPs to access webpages.
  - They, by multiple IPs and without an identifiable UA, just behave like normal real users. This means that **we can't do anything about them from MyBB's perspective at the statistics module**.
- Human users accessing behind a NAT that has a very few public Internet IPs are really really real users.
  - Yes, from the forum's perspective, they just behave like bots. If we don't distinguish them from the above two types of accesses, we're just blinding ourselves.

However, I do agree to your idea of a meaningful _most users online_, so there's a further thought:
1. Change the name `The most users online at one time` to `The most accesses/visits at one time`;
2. Add another variable of `The most registered user online at one time` which IMO is much meaningful than its predecessor in some circumstances.
